x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java

```diff
@@ -0,0 +1,194 @@
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        // If this node is not a master node, exit.
+        if (event.state().nodes().isLocalNodeElectedMaster() == false) {
+            return;
+        }
+
+        if (event.state().nodes().getMaxNodeVersion().after(event.state().nodes().getSmallestNonClientNodeVersion())) {
+            logger.debug("Skipping up-to-date check as cluster has mixed versions");
+            return;
+        }
+
+        // ensure that index templates are present
+        if (isAllTemplatesCreated(event) == false) {
+            logger.trace("Skipping index creation; not all templates are present yet");
+            return;
+        }
+
+        addIndicesIfMissing(event.state());
+    }
+
+    protected boolean isAllTemplatesCreated(ClusterChangedEvent event) {
+        return ProfilingIndexTemplateRegistry.isAllTemplatesCreated(event.state());
+    }
+
+    private void addIndicesIfMissing(ClusterState state) {
+        Optional<Map<String, IndexMetadata>> maybeMeta = Optional.ofNullable(state.metadata().indices());
+        for (Map.Entry<String, String> idxAlias : INDICES_AND_ALIASES.entrySet()) {
+            String index = idxAlias.getKey();
+            String alias = idxAlias.getValue();
+            final AtomicBoolean creationCheck = indexCreationInProgress.computeIfAbsent(index, key -> new AtomicBoolean(false));
+            if (creationCheck.compareAndSet(false, true)) {
+                final boolean indexNeedsToBeCreated = maybeMeta.flatMap(idxMeta -> Optional.ofNullable(idxMeta.get(index)))
+                    .isPresent() == false;
+                if (indexNeedsToBeCreated) {
+                    logger.debug("adding index [{}], because it doesn't exist", index);
+                    putIndex(index, alias, creationCheck);
+                } else {
+                    logger.trace("not adding index [{}], because it already exists", index);
+                    creationCheck.set(false);
+                }
+            }
+        }
+    }
+
+    private void onPutIndexFailure(String index, Exception ex) {
+        logger.error(() -> format("error adding index [%s] for [%s]", index, ClientHelper.PROFILING_ORIGIN), ex);
+    }
+
+    private void putIndex(final String index, final String alias, final AtomicBoolean creationCheck) {
+        final Executor executor = threadPool.generic();
+        executor.execute(() -> {
+            CreateIndexRequest request = new CreateIndexRequest(index);
+            if (alias != null) {
+                try {
+                    Map<String, Object> sourceAsMap = Map.of("aliases", Map.of(alias, Map.of("is_write_index", true)));
+                    request.source(sourceAsMap, LoggingDeprecationHandler.INSTANCE);
+                } catch (Exception ex) {
+                    creationCheck.set(false);
+                    onPutIndexFailure(index, ex);
```

Comment:
I am trying to think whether there are other cases that could throw an exception and we need to release the flag. I haven't found one yet but I put it here in case it triggers you or someone else.

Reply:
I did notice that I missed adding a `return` here and added it in https://github.com/elastic/elasticsearch/pull/95666/commits/f1dee21951b7bae18cd687618df69e5cd7987cd1.